File: backend/src/main/java/com/eduardo/backend/services/CompanyService.java
package com.eduardo.backend.services;

import com.eduardo.backend.dtos.CompanyApproveDTO;
import com.eduardo.backend.dtos.CompanyCreateDTO;
import com.eduardo.backend.dtos.CompanyResponseDTO;

import java.util.List;

/**
 * Contrato das regras de negócio
 * relacionadas às empresas.
 */
public interface CompanyService {

    // Cria uma nova empresa
    CompanyResponseDTO createCompany(CompanyCreateDTO dto);

    // Atualiza dados da empresa
    CompanyResponseDTO updateCompany(Long companyId, CompanyCreateDTO dto);

    // Lista empresas pendentes (ADMIN)
    List<CompanyResponseDTO> getPendingCompanies();

    // Aprova ou rejeita empresa
    CompanyResponseDTO approveOrRejectCompany(Long companyId, CompanyApproveDTO dto);

    // Lista empresas do OWNER
    List<CompanyResponseDTO> getCompaniesByOwner();

    // Busca empresa por ID
    CompanyResponseDTO getCompany(Long companyId);
}

File: frontend/src/app/core/services/company.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { CompanyDTO } from '../models/company.model';

@Injectable({ providedIn: 'root' })
export class CompanyService {
  private apiUrl = 'http://localhost:8080/api/companies';

  constructor(private http: HttpClient) {}

  // OWNER cria empresa → retorna CompanyResponseDTO
  createCompany(data: CompanyDTO): Observable<CompanyDTO> {
    return this.http.post<CompanyDTO>(`${this.apiUrl}`, data);
  }

  // OWNER ATUALIZA EMPRESA JÁ EXISTENTE
  updateCompany(id: number, data: CompanyDTO): Observable<CompanyDTO> {
    return this.http.put<CompanyDTO>(`${this.apiUrl}/${id}`, data);
  }

  // ADMIN busca pendentes
  getPendingCompanies(): Observable<CompanyDTO[]> {
    return this.http.get<CompanyDTO[]>(`${this.apiUrl}/pending`);
  }

  // ADMIN aprova/reprova
  approveCompany(companyId: number, approved: boolean): Observable<CompanyDTO> {
    return this.http.put<CompanyDTO>(
      `${this.apiUrl}/${companyId}/approve`,
      { approved } // CompanyApproveDTO
    );
  }

  // OWNER busca empresas dele
  getCompaniesByOwner(): Observable<CompanyDTO[]> {
    return this.http.get<CompanyDTO[]>(`${this.apiUrl}/mine`);
  }

  // qualquer um pega empresa por id
  getCompany(id: number): Observable<CompanyDTO> {
    return this.http.get<CompanyDTO>(`${this.apiUrl}/${id}`);
  }
}
